### ISSUE TITLE
Enhance writeWithUnsafeMutableBytes

### ISSUE THREAD

**Issue by Lukasa** [apple/swift-nio#1105]
In general, we don't want people using the unsafe APIs on `ByteBuffer`, but in the cases where it really is the best solution we should ensure that it works as well as it reasonably can.

Most of this API is pretty ok, but `writeWithUnsafeMutableBytes` is a very tricky function to use properly. This is because, unlike all of the other `write` functions, it is not able to tolerate arbitrarily large writes. The reason here is that by the time the body gets the pointer to the writable bytes it is not possible to resize the underlying buffer.

The safe way to use this function is to preflight by telling the `ByteBuffer` how many bytes you need. You can do this via `reserveCapacity`, but calculating the correct number to pass to that function is annoying.

We should enhance `writeWithUnsafeMutableBytes` to take an `Int` that indicates how many bytes the caller would like to be free, and then we can do this calculation ourselves.

**Comment by 2bjake** [comment]
I thought I'd take a stab at this one, but I wanted to make sure I understood the intention correctly. Does this look right?

```swift
    public mutating func writeWithUnsafeMutableBytes(bodyByteCount: Int? = nil, _ body: (UnsafeMutableRawBufferPointer) throws -> Int) rethrows -> Int {
        if let bodyByteCount = bodyByteCount, bodyByteCount > capacity - writerIndex {
            reserveCapacity(writerIndex + bodyByteCount)
        }
        // write bytes (existing function body)
    }
```

**Comment by Lukasa** [comment]
Broadly right, but we want to do a bit more. Specifically we want the user to ask for a minimum number of uninitialised bytes, not a total capacity, then calculate what the required capacity is.

**Comment by 2bjake** [comment]
The user should be able to specify the minimum number of uninitialized bytes (writableBytes?) before the write occurs? If that's the case, I think this code does that (though the name ```bodyByteCount``` should be changed to something like ```withReservedWritableBytes```).

Unless I'm mistaken (which is quite possible!) adding the user specified writableBytes count to the ```writerIndex``` will calculate a new minimumCapacity that can be passed to ```reserveCapacity```.

**Comment by Lukasa** [comment]
Yeah, writable bytes, and good catch: I missed that.  That’s what I get for reviewing while walking down the street!

**Comment by Lukasa** [comment]
So yes, this is broadly correct.

**Comment by weissi** [comment]
fixed with #1175
